`builtin -d` should not delete special builtins

JohnoKing opened the pull request:
The man page for the `builtin` command says special builtins cannot be deleted with `builtin -d`:
>  -d              Deletes each of the specified built-ins. Special built-ins cannot be deleted.

This flag doesn't behave as documented, running `builtin -d` on a special builtin does delete it. As an example, the following set of commands ends with 'export: not found':
```
$ builtin -d export
$ export foo=bar
```
This pull request backports the bugfix from ksh93v- (2014-12-24-beta), which added an error check to prevent special builtins from being deleted.

McDutchie commented:
Many thanks for your pull request, @JohnoKing !

Yes, that is definitely a bug, thanks for finding it. Unfortunately I am having trouble understanding the fix (or the whole emacs handling code for that matter).

Scanning through the code, though, I find that the -2 value of `ep->prevdirection` has a special meaning in various places, as it is specifically tested for — whereas the -3 value that you introduce is new, and does not seem to be handled anywhere. Doesn't this risk introducing other bugs?

After your fix, it never gets assigned a value of -2 under any circumstances, so even if that is valid, all those tests for `ep->prevdirection == -2` are now pointless and should be removed. I suspect that can't be right.

You've explained the *what* and the *how* of this fix, but can you explain *why* this fixes it?

JohnoKing commented:
This bug was occuring because ksh was only doing a reverse search
based on the first command that was completed using the up arrow.
All subsequent commands were ignored as ksh was saving the first
command and only based later searches off of it.

I have revised my pull request to use the -2 value for `ep->prevdirection` and
added better information for reproducing this bug.

McDutchie commented:
Thanks, that looks a great deal more sensible. I'm ready to merge that.

I'm not going to ask you to jump through hoops, but for future reference: I really don't need the rationale and reproducer in the NEWS file; that is meant to be a concise overview of significant changes (although I am still a great deal more verbose than AT&T's original changelogs!). I'd like the full details in commit messages instead.

In fact, your adding the details to the NEWS file did just inspire me to link it to github's version of `git blame NEWS` from README.md, so that people can easily click on the associated commit messages to get the full nitty-gritty.

JohnoKing committed:
`builtin -d` should not delete special builtins

The man page for the builtin command says special builtins cannot
be deleted. This wasn't the case though, running `builtin -d` on
a special builtin was deleting it. As an example, the following
set of commands was ending with 'export: not found':

$ builtin -d export
$ export foo=bar

This commit backports the bugfix from ksh93v- (2014-12-24-beta),
which added an error check to prevent special builtins from being
deleted.

src/cmd/ksh93/sh/nvdisc.c:
 - Add an error check to prevent special builtins from being deleted.

src/cmd/ksh93/tests/builtins.sh
 - Add a regression test for using `builtin -d` on special builtins.

McDutchie commented on src/cmd/ksh93/tests/builtins.sh:
```diff
@@ -686,5 +686,10 @@ PATH=$tmp:$PATH $SHELL <<-\EOF || err_exit "'whence' gets wrong path on init"
+# `builtin -d` should not delete special builtins
+builtin -d export 2> /dev/null
+whence -q export || err_exit '`builtin -d` deletes special builtins'
+
```
Deleting `export` in the main shell environment will affect subsequent tests that may be added. I would prefer it if the test were non-destructive by executing `builtin -d` plus the check in a subshell:

```sh
(builtin -d export 2> /dev/null
whence -q export) || err_exit '`builtin -d` deletes special builtins'
```

JohnoKing replied:
Thanks for pointing that out, it is now fixed.

McDutchie replied:
…I just thought of another thing, sorry. An external `export` command may exist on some systems. So to avoid a false positive, we need to disable the PATH search in the subshell.
```sh
(builtin -d export 2> /dev/null
PATH=/dev/null
whence -q export) || err_exit '`builtin -d` deletes special builtins'
```

JohnoKing replied:
That is now fixed as well.

McDutchie reviewed (changes requested):
Thank you (again)! Just one little thing...
